### DISCUSSION TITLE
[how to remove cropping] Is resolution=1024 required? Can I not train a LORA with Dreambooth for SDXL with 1920x1080 images?

### DISCUSSION THREAD

**Discussion by dkebudi** [huggingface/diffusers#9018] [Q&A] [answered]
I would ideally want the output to be the same aspect ratio. Thanks everyone!

**Comment by asomoza** [comment]
Hi, the bests results with SDXL will be with the same resolutions as it was trained with, which are in the [paper](https://arxiv.org/abs/2307.01952):

![image](https://github.com/user-attachments/assets/d0f747b4-9f9d-42a1-9cb8-0af3c4cf2baa)

Since you're training a Dreambooth LoRA, this will probably won't have any impact though, DB LoRAs are really small and practically it doesn't matter the resolution you use if it's not lower or a bad quality one, this will only have an effect for finetunes. 

With SDXL I can still generate an image like this one without a problem:

![20240730171610](https://github.com/user-attachments/assets/ddd456ef-1c99-40ed-8740-e9e485ff5ab0)

For example I did an experiment when SDXL came out with this [dataset](https://huggingface.co/datasets/OzzyGT/melvyn) which is just 1024px square images.

And with that training I could still generate an extreme landscape image like this one without any problems.

![image](https://github.com/user-attachments/assets/ca2ea3c3-0c31-4edf-bbdf-f44d63319649)

For DB LoRAs is a lot more important the quality and description of what you're training than the aspect ratio of the images which the model probably won't learn with just a few images.

**Comment by dkebudi** [comment]
@asomoza thank you! I am using [this script](https://github.com/huggingface/diffusers/blob/main/examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py#L48) and it seems like it forces the model to be 1024x1024. I want to force my own aspect ratio. 

With the [UI HuggingFace generated ](https://huggingface.co/spaces/multimodalart/lora-ease) I was hoping to be able to make enough tweaks to the under lying code such that I allow for different aspect ratios in trianing. 

Currently, the way the infra is set up, it center crops my data to fit a square with padding, in this instance 1920x1920. 

Any ideas? 

Any help would be appreciated.

**Reply by asomoza** [reply, marked as answer]
That's a two part question, the script has some arguments to set the resolution and cropping of the images:

`--resolution` and `--center_crop` which seems you're using.

I don't really use the default scripts because I'm always testing new stuff, but if you want to change it so it can accept non-square images you will need to change the [Dataset](https://github.com/huggingface/diffusers/blob/ea1b4ea7ca5fa7ab3d2094a70a35481f2a1036c2/examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py#L979) object. 

[This line](https://github.com/huggingface/diffusers/blob/ea1b4ea7ca5fa7ab3d2094a70a35481f2a1036c2/examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py#L1073) is the one that does the resizing so you'll need to change it. If you're using the same resolution images you can practically just delete the resizing from the dataset.

> make enough tweaks to the under lying code such that I allow for different aspect ratios in trianing

Just as a heads up, If you're struggling with just the image resizing the "different aspect ratios training" will not be easy for you, this involves some more complex logic where you will need to use bucketing and know how to resolve when the batches don't match, so you'll practically need to code your own dataset object and how to retrieve the images from it.

**Comment by dkebudi** [comment]
This is really helpful, thank you @asomoza. 

Lastly -- given the Lora-Ease wrapper I am using, is there a way for me to manipulate the Dataset object that code calls? Asking since Lora-Ease is also Hugging Face owned. A path forward would be phenomenal. 

I will exclusively train for 1920x1080 for the time being.

**Reply by asomoza** [reply]
I suppose you mean to be able to use it in spaces and not locally.

Since that space is running in `zero` and works, you can just clone the repo and use it as your own personal space. You can upload to the same repo the modifidied training script and change this [line](https://huggingface.co/spaces/multimodalart/lora-ease/blob/60c7a9e56190030351fedb2fb942b8712ad17fff/app.py#L25) to load your own custom training script in which you can change the resize part so it doesn't change the aspect ratio or the resolution.

**Reply by dkebudi** [reply]
you are the best. almost there. 

do you recommend that I just take out the resize line, and then delete any references to it in the code?

**Reply by asomoza** [reply]
That would accomplish what you want but I don't recommend it, this will require more VRAM (not sure if zero will be able to run it) and also it will make the training slower with little to no benefit, but yeah, you can delete the Resize transformation and the references to it. 

Also I recommend to do a training with manual resizing and cropping to 1024px and see if there's a real difference, I say manual so you can ensure that the image contains what you're training for.

**Reply by dkebudi** [reply]
so your recommendation is just to hard code the aspect ratio in instead? 

I was also planning on taking (edit) **out any reference to crop.

**Reply by asomoza** [reply]
yeah, for DB LoRA you're fine just cropping, the only problem I encounter is that when you do automatic cropping and text descriptions (specially with augmentations and repeats) is that what the image doesn't contain the description and thus makes the training really bad.

So for example if you're training a woman and the augmentation crops the head but the description says "blonde woman", you're pretty much making the model learn that `blonde` its a random thing from the rest of the image.

That's why the best dataset is the one you make youself manually with manual cropping and manual descriptions and that probably won't change in the near future.

You can achieve something really close if you use something like Florence to detect the subject, crop the image around it and then make the text description, also since SDXL has microconditioning you can tell it that the image was cropped and the original position of the subject.

Edit: Since I though this was about cropping I will still leave it here but it doesn't apply to your question after the edit.

**Reply by asomoza** [reply]
I'm also suggesting you do the experiment yourself so you can see for yourself, I still don't know why do you want (at all costs) to train a DB LoRA in a different resolution.

I sometime ago saw some youtubers and influencers talk about that it does make a difference but at the same time they recommend to train with a rank of 128/256 with a high learning rate which is really bad because it produces a 1.5 GB overfitted lora which is good to show their face in a generation of a standing person but doesn't work for anything else.

This is not sure science, you can even get different results with the same training params, so you'll need to experiment a lot. I also think there's a benefit in training with multiple aspect ratios but just not for small datasets and dreambooth to be worth the trouble.

**Reply by dkebudi** [reply]
great insights on your first comment. I am new to the SDXL world, but have a pretty decent background in coding -- will work with the information you presented to see if I can dive deeper into augmentation and microconditioning. 

thank you so much @asomoza !

The training data I have is in 1920x1080 and the edges of the image have information I want to preserve. So trying to fit the training to the size I am working with. 

Meanwhile I am getting an odd error following your recommendation: 

```
Starting training...
  File "/app/073024-test3/trainer.py", line 1134
    else:
    ^
IndentationError: expected an indented block after 'if' statement on line 1130
Traceback (most recent call last):
  File "/app/073024-test3/script.py", line 129, in <module>
    main()
  File "/app/073024-test3/script.py", line 123, in main
    do_train(script_args)
  File "/app/073024-test3/script.py", line 26, in do_train
    raise Exception("Training failed.")
Exception: Training failed.
```

any insights?

**Reply by dkebudi** [reply]
could it be because of how I am referencing the training script is the custom repo:

`training_script_url = "https://huggingface.co/spaces/dkebudi/lora/raw/main/train_dreambooth_lora_sdxl_advanced.py"
`

as opposed to the url structure as in the main lora-ease repo:

`
training_script_url = "https://raw.githubusercontent.com/huggingface/diffusers/ba28006f8b2a0f7ec3b6784695790422b4f80a97/examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py"`

**Reply by asomoza** [reply]
if the file is visible and a text file it should work but if the repo is private the space won't be able to load it. For example if I try to load `https://huggingface.co/spaces/dkebudi/lora/raw/main/train_dreambooth_lora_sdxl_advanced.py` I get a `Repository not found`

**Reply by asomoza** [reply]
> The training data I have is in 1920x1080 and the edges of the image have information I want to preserve.

This is pretty much the only real use case where it's worth going the extra mile to train with the full image in the original aspect ratio but personally I haven't encountered this specific scenario yet so I don't have any real experience or results with this kind of training.

**Reply by dkebudi** [reply]
ah makes sense. Let me try to make the repo non-private.

**Reply by dkebudi** [reply]
Still getting this guy:
```

Starting training...
  File "/app/073024-test-repo-v4/trainer.py", line 1134
    else:
    ^
IndentationError: expected an indented block after 'if' statement on line 1130
Traceback (most recent call last):
  File "/app/073024-test-repo-v4/script.py", line 129, in <module>
    main()
  File "/app/073024-test-repo-v4/script.py", line 123, in main
    do_train(script_args)
  File "/app/073024-test-repo-v4/script.py", line 26, in do_train
    raise Exception("Training failed.")
Exception: Training failed.
```

**Reply by dkebudi** [reply]
tried with another repo: 
`
https://huggingface.co/spaces/dkebudi/lora-ease-dk/raw/main/train_dreambooth_lora_sdxl_advanced.py
`

**Reply by asomoza** [reply]
oh, then it's probably the indent the error its saying, that if...else statement is not correctly indented.

If it is the same script you shared: https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/main/train_dreambooth_lora_sdxl_advanced.py#L1130

you're commenting all the code inside the if..else so you can safely delete it or if you don't want to , you can do this:

```pyhon
                if args.center_crop:
                    ...
                    #y1 = max(0, int(round((image.height - args.resolution) / 2.0)))
                    #x1 = max(0, int(round((image.width - args.resolution) / 2.0)))
                    #image = train_crop(image)
                else:
                    ...
                    #y1, x1, h, w = train_crop.get_params(image, (args.resolution, args.resolution))
                    #image = crop(image, y1, x1, h, w)
```

**Reply by dkebudi** [reply]
Let me check!

**Reply by dkebudi** [reply]
that worked for the if statement -- thank you @asomoza !

But now I am getting:

```
Traceback (most recent call last):
  File "/app/autotrain-073024-test-repo-v5/trainer.py", line 54, in <module>
    from diffusers import (
ImportError: cannot import name 'EDMEulerScheduler' from 'diffusers' (/app/env/lib/python3.10/site-packages/diffusers/__init__.py)
Traceback (most recent call last):
  File "/app/autotrain-073024-test-repo-v5/script.py", line 129, in <module>
    main()
  File "/app/autotrain-073024-test-repo-v5/script.py", line 123, in main
    do_train(script_args)
  File "/app/autotrain-073024-test-repo-v5/script.py", line 26, in do_train
    raise Exception("Training failed.")
Exception: Training failed.
```

**Reply by dkebudi** [reply]
is there something I need to do regarding the env and it's dependencies? Now that I am thinking, I probably need to bring in the requirements.txt from the diffusers repo into my repo right?

**Reply by asomoza** [reply]
yeah

**Reply by dkebudi** [reply]
<img width="910" alt="Screenshot 2024-07-30 at 9 41 51 PM" src="https://github.com/user-attachments/assets/935873fb-1141-4ecb-8014-020416d50830">

@asomoza  Still getting the same error that it can't find EDMEulerScheduler. My guess is I need to add the packages to the above part of the code in [app.py](https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/main/app.py)

Any thoughts?

**Reply by asomoza** [reply]
This part:

`git+https://github.com/huggingface/diffusers@ba28006f8b2a0f7ec3b6784695790422b4f80a97`

change it to just use the main:

`git+https://github.com/huggingface/diffusers`

I'm not that familiar with that space so I can't really help you a lot, it seems that it switches the GPU and does a second installation. Basically to train a LoRA, you don't need specific versions of the repos, it should work with just the library names without specific versions.

**Reply by dkebudi** [reply]
Thank you @asomoza ! 

The main issue I am facing right now is that the training script I am copying over to my own repo has a lot of dependencies in the diffusers repo. Since I am only copying the training script it is failing to find the necessary files:
```

Steps:   0%|          | 0/600 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/app/autotrain-dkebudi-073024-test-repo-v6/trainer.py", line 2451, in <module>
    main(args)
  File "/app/autotrain-dkebudi-073024-test-repo-v6/trainer.py", line 2074, in main
    model_input = model_input * vae.config.scaling_factor
UnboundLocalError: local variable 'vae' referenced before assignment
Steps:   0%|          | 0/600 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/app/autotrain-dkebudi-073024-test-repo-v6/script.py", line 129, in <module>
    main()
  File "/app/autotrain-dkebudi-073024-test-repo-v6/script.py", line 123, in main
    do_train(script_args)
  File "/app/autotrain-dkebudi-073024-test-repo-v6/script.py", line 26, in do_train
    raise Exception("Training failed.")
Exception: Training failed.
```

Any advice on best practices here? Ideally I just clone the whole repo to my repo or vice versa. Thank you!

**Reply by dkebudi** [reply]
@asomoza 

So I realized -- I am passing in the path for a vae for the AutoEncoder to load. Thus sub_folder doesn't matter in the script. As such, there should be no issue loading up the vae in main(). Any ideas here?

**Reply by asomoza** [reply]
I think you're good with the dependencies now, I don't really know what's your code doing but its clear that it doesn't find the vae, so double check that the vae has the correct path, that the space can actually see it and that it's loading it, you can just do a print somewhere after the loading to check if the vae is None or not.

**Reply by dkebudi** [reply]
yea found it ... bug with the repo I forked

It has this line I commented out, which deletes vae from memory if the validation prompt is null.
<img width="609" alt="Screenshot 2024-07-31 at 4 39 03 PM" src="https://github.com/user-attachments/assets/6fec4301-fef9-4219-986d-02eb3be0a024">

HOWEVER 

if the latents mean and sdt is null at the same time, it uses VAE to calculate model input:

<img width="757" alt="Screenshot 2024-07-31 at 4 39 54 PM" src="https://github.com/user-attachments/assets/2407c2f1-4f3e-4ad2-9ff2-4e78fd22fd85">

which errors out. 

I think the reason I am getting latents and sdt null is bc of the diffusers repo version I am using. 

But never the less, this is likely an edge case we should account for in the repo.

Thanks @asomoza !

**Reply by dkebudi** [reply]
It is training with the del vae commented out. I will also try to train with the old diffusers repo version to see if the latents are still null.

**Reply by dkebudi** [reply]
@asomoza I am still getting squares as outputs

here is the updated script: https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/main/train_dreambooth_lora_sdxl_advanced.py

any ideas what would be causing this?

**Reply by dkebudi** [reply]
Can't figure it out for the life of me >.<

everything seems to be hard coded to 1080 (h) x 1920 (w) -- how could it produce anything in validation that is outside of those pre-sets? I think I am missing something.

**Reply by asomoza** [reply]
this is just my guess, but the validation prompt doesn't pass the height and width so it defaults to 1024x1024

https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/main/train_dreambooth_lora_sdxl_advanced.py#L2413

**Reply by dkebudi** [reply]
also tagging @linoytsaban here in case they have any tips!

**Reply by dkebudi** [reply]
hmm how would i pass in the dimensions in the validation prompt? @asomoza

**Reply by asomoza** [reply]
pipeline_args = {"prompt": args.validation_prompt, "num_inference_steps": 25, "width":1920, "height":1080}

**Reply by dkebudi** [reply]
amazing! thanks. 

should i do it here as well:

https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/439be089987dd1c464a7bfcb016b655e5d4d6ffc/train_dreambooth_lora_sdxl_advanced.py#L2349

**Reply by dkebudi** [reply]
also thoughts on the VAE Null latents mean and sdt, do you think my assumption that it is caused by the diffusers version is valid?

**Reply by asomoza** [reply]
> should i do it here as well

yeah

> also thoughts on the VAE Null latents mean and sdt, do you think my assumption that it is caused by the diffusers version is valid?

Probably, that space had a specific version of diffusers harcoded which means probably @apolinario was working with it while waiting for some changes in the main repo. I don't really use that training script so @apolinario and @linoytsaban can answer this question better.

**Reply by dkebudi** [reply]
roger that -- you have saved my life in the past two days. Much appreciated. 

@apolinario + @linoytsaban if you guys have any insights into how to do this the right way / first class way -- I would really appreciate it!

**Reply by dkebudi** [reply]
@asomoza the output on the model card is still a square ... hmm. What other rocks have I not turned over yet? could it be anything else in the pipeline? the VAE?

**Reply by asomoza** [reply]
what about the validation images? can you view the validation in the tensorboard?

You can also try to set the width and height directly here:

https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/main/train_dreambooth_lora_sdxl_advanced.py#L281

```python
[pipeline(**pipeline_args, width=1920, height=1080, generator=generator).images[0]
```

but it should be the same.

**Reply by dkebudi** [reply]
<img width="1381" alt="Screenshot 2024-07-31 at 6 59 50 PM" src="https://github.com/user-attachments/assets/fa9ed4ce-8767-4226-af42-3b1c6b876819">

Great q on tensorboard. I haven't been logging images there -- at leat if I am navigating it correctly, doesn't seem like I have been logging them. Do you know how I can? 

Yeah adding the args there should be the same effectively.

**Reply by asomoza** [reply]
they should be added automatically, if you don't choose a tracker, tensorboard its the default and also it should add the images here:

https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/main/train_dreambooth_lora_sdxl_advanced.py#L287

You can also manually save them and check their resolution, those images should be the same that are uploaded to the model card (the final inference ones).

**Reply by dkebudi** [reply]
So basically they will be 1024x1024? That is the dimensions of the images I am seeing in the model card.

**Reply by asomoza** [reply]
they shouldn't be, that's why asking you to check the images generated, to discard that maybe they're being resized somewhere else. If the images generated and used in the validation are correct, then they probably are being changed in the code that uploads them to the hub.

**Reply by dkebudi** [reply]
I am getting confused here. I get your logic -- and I want to carry it out. What is the best way for me to test right now? Download the model, put it into Comfy, generate images, and see what I get? How would that validate that the model was trained on 1920x1080 since I am defining the empty latent image?

If I download the images from the model card's landing page, which I believe are the validation images, the images are 1024x1024. If I generate images using the hosted model on the model card, the output is also 1024x1024. Tensorboard doesn't seem to hold my images, unless I don't know how to get to the images saved. 

@asomoza

**Reply by asomoza** [reply]
that is a different issue, I was trying to help you see why the model images in the model card were in a square ratio because it seemed important to you.

You probably won't be able to validate that the model was trained with 1920x1080 since SDXL can generate in that resolution just fine as I showed before, also with a small training and specially with a DB LoRA the model won't learn a new resolution as I told you before.

The generated resolution is not important here, you just need to check that it learned what you wanted it to learn with the training. You can use any resolution but if the online generation is hard coded to 1024x1024, it will be better to just download the model and test it locally.

**Reply by dkebudi** [reply]
thanks @asomoza -- to clarify: my goal is to make sure that the model is not cropping out important details on the edges of the training images, as I presumed it was doing, when training on 1024x1024. I wanted to swap the training data with a 1920x1080p so that it would capture the entirety of the image. My dataset will have >1000 images, and even >5000 images to train with. 

When it comes to inference, once I validate that the training is indeed happening as it should, it's straight forward for me to output images in any size I want as you said. I will just work with my empty latent image when initializing. 

So -- what do you recommend is the best way to validate that? any print statements I can use that you recommend? or tensorboard? not sure how relevant tensorboard is after the clarification.

**Reply by asomoza** [reply]
For image validation in the training you can use tensorboard, wandb or just save the image after each epoch or number of steps, at the end, its your choice, locally I can use any of them and they work but I don't know why in your training the tensorboard doesn't have any images.

I suggest you use what is the easiest and faster for you, also as you said, you don't need to generate in a specific resolution, just check that all the details you want are present in the generation. Still don't know on what you're training, but if you have small details at the edges of the image, you probably will need those 5000 images and to train for a long time.

**Reply by asomoza** [reply]
Oh and one small detail, to make sure it's using the full images and not cropping them, check the VRAM consumption, training in 1920x1080 should use more VRAM than 1024x1024.

**Reply by dkebudi** [reply]
Gotcha on the long training and using all the images.

Yeah weird that tensor-board is not working. Primarily because I have changed the code to force pushing images to TB in every run. Nevertheless, anything easy that comes to mind right now? Where in the code can I check that the images going into the model for training are 1920x1080? I will just add a print statement.

**Reply by asomoza** [reply]
not 100% sure in that script but you should be able to print the shape after this line if you're not caching the latents:

https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/main/train_dreambooth_lora_sdxl_advanced.py#L2070

```python
print(f"{pixel_values.shape=}")
```

The last two values should be the height and width.

**Reply by dkebudi** [reply]
Awesome. 

I opted to put it here since I am caching: https://huggingface.co/spaces/dkebudi/lora-ease-dk/blob/3c525b4aecf4fa3eecd273c5bbb7bb451fa1a169/train_dreambooth_lora_sdxl_advanced.py#L1896

And it is telling me that the images are 1080x1920! 

<img width="481" alt="Screenshot 2024-07-31 at 10 50 21 PM" src="https://github.com/user-attachments/assets/4131f448-ed6e-45b3-95cb-05a73cf1bc5e">

Thank you @asomoza !

**Reply by ScilenceForest** [reply]
@asomoza Hello asomoza, I would like to ask for the training of SDXL Controlnet, the size of the image that I need to generate is 2048*2048, and my dataset image is also 2048*2048, but I am training Controlnet at 1024*1024 size (resize the dataset image), and I would like to know in this scenario Should I train Controlnet at 2048*2048? sorry this is my first reply, I'm not sure if this question needs to start a new discussion, I'm sorry if it's bothering you or the author of the original question (and please make me aware of it)

**Reply by asomoza** [reply]
Hi, training with 2048 probably won't give you any additional benefit, even though SDXL can generate at 2048x2048, most of the time it doesn't do it good and training a controlnet won't fix that. Your best option would be to train in 1024px and also generate at 1024px and then do an upscale.

If you need further help it would be best to start a new discussion, mostly so people and maybe even yourself in the future can find the answers and information better.
